PowerEditor/src/ScitillaComponent/Buffer.cpp

```diff
@@ -313,44 +313,72 @@ void Buffer::reload()
 	}
 }
 
-int Buffer::getFileLength() const
+int64_t Buffer::getFileLength() const
 {
 	if (_currentStatus == DOC_UNNAMED)
 		return -1;
 
-	struct _stat buf;
-
-	if (PathFileExists(_fullPathName.c_str()))
+	WIN32_FILE_ATTRIBUTE_DATA attributes;
+	if (GetFileAttributesEx(_fullPathName.c_str(), GetFileExInfoStandard, &attributes) != 0)
 	{
-		if (!generic_stat(_fullPathName.c_str(), &buf))
-			return buf.st_size;
+		LARGE_INTEGER size;
+		size.LowPart = attributes.nFileSizeLow;
+		size.HighPart = attributes.nFileSizeHigh;
+		return size.QuadPart;
 	}
 	return -1;
 }
 
 
 generic_string Buffer::getFileTime(fileTimeType ftt) const
 {
-	if (_currentStatus == DOC_UNNAMED)
-		return generic_string();
-
-	struct _stat buf;
+	generic_string result;
 
-	if (PathFileExists(_fullPathName.c_str()))
+	if (_currentStatus != DOC_UNNAMED)
 	{
-		if (!generic_stat(_fullPathName.c_str(), &buf))
+		WIN32_FILE_ATTRIBUTE_DATA attributes;
+		if (GetFileAttributesEx(_fullPathName.c_str(), GetFileExInfoStandard, &attributes) != 0)
 		{
-			time_t rawtime = (ftt == ft_created ? buf.st_ctime : (ftt == ft_modified ? buf.st_mtime : buf.st_atime));
-			tm *timeinfo = localtime(&rawtime);
-			const int temBufLen = 64;
-			TCHAR tmpbuf[temBufLen];
+			FILETIME rawtime;
+			switch (ftt)
+			{
+			case ft_created:
+				rawtime = attributes.ftCreationTime;
+				break;
+			case ft_modified:
+				rawtime = attributes.ftLastWriteTime;
+				break;
+			default:
+				rawtime = attributes.ftLastAccessTime;
+				break;
+			}
+
+			SYSTEMTIME utcSystemTime, localSystemTime;
+			FileTimeToSystemTime(&rawtime, &utcSystemTime);

```

Comment:
Why this conversion? Does FILETIME is already UTC?
`Contains a 64-bit value representing the number of 100-nanosecond intervals since January 1, 1601 (UTC).`
https://msdn.microsoft.com/en-us/library/windows/desktop/ms724284(v=vs.85).aspx

Reply:
Yes, but `getFileTime` returns a string for file statistics, so it should be presented to the user in his local time, not UTC. Therefore, `SystemTimeToTzSpecificLocalTime` needs to be used, and since it operates on `SYSTEMTIME`, first this file time needs to be converted to system time.